x-pack/plugins/lens/public/app_plugin/app.tsx

```diff
@@ -220,379 +198,410 @@ export function App({
-      .save(doc)
-      .then(({ id }) => {
-        core.chrome.recentlyAccessed.add(getFullPath(id), doc.title, id);
-        // Prevents unnecessary network request and disables save button
-        const newDoc = { ...doc, id };
-        const currentOriginatingApp = state.originatingApp;
+      if (
+        attributeService.inputIsRefType(newInput) &&
+        newInput.savedObjectId !== originalSavedObjectId
+      ) {
+        chrome.recentlyAccessed.add(
+          getFullPath(newInput.savedObjectId),
+          docToSave.title,
+          newInput.savedObjectId
+        );
         setState((s) => ({
           ...s,
           isSaveModalVisible: false,
-          originatingApp:
-            newlyCreated && !saveProps.returnToOrigin ? undefined : currentOriginatingApp,
-          persistedDoc: newDoc,
-          lastKnownDoc: newDoc,
+          isLinkedToOriginatingApp: false,
         }));
-        if (docId !== id || saveProps.returnToOrigin) {
-          redirectTo(id, saveProps.returnToOrigin, newlyCreated);
-        }
-      })
-      .catch((e) => {
-        // eslint-disable-next-line no-console
-        console.dir(e);
-        trackUiEvent('save_failed');
-        core.notifications.toasts.addDanger(
-          i18n.translate('xpack.lens.app.docSavingError', {
-            defaultMessage: 'Error saving document',
-          })
-        );
-        setState((s) => ({ ...s, isSaveModalVisible: false }));
-      });
-  };
+        redirectTo(newInput.savedObjectId);
+        return;
+      }
 
-  const onError = useCallback(
-    (e: { message: string }) =>
-      core.notifications.toasts.addDanger({
-        title: e.message,
-      }),
-    [core.notifications.toasts]
-  );
+      const newDoc = {
+        ...docToSave,
+        ...newInput,
+      };
+      setState((s) => ({
+        ...s,
+        persistedDoc: newDoc,
+        lastKnownDoc: newDoc,
+        isSaveModalVisible: false,
+        isLinkedToOriginatingApp: false,
+      }));
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.dir(e);
+      trackUiEvent('save_failed');
+      notifications.toasts.addDanger(
+        i18n.translate('xpack.lens.app.docSavingError', {
+          defaultMessage: 'Error saving document',
+        })
+      );
+      setState((s) => ({ ...s, isSaveModalVisible: false }));
+    }
+  };
 
   const { TopNavMenu } = navigation.ui;
 
+  const savingPermitted = Boolean(state.isSaveable && application.capabilities.visualize.save);
+  const topNavConfig = getLensTopNavConfig({
+    showSaveAndReturn: Boolean(
+      state.isLinkedToOriginatingApp &&
+        // Temporarily required until the 'by value' paradigm is default.
+        (dashboardFeatureFlag.allowByValueEmbeddables || Boolean(initialInput))
+    ),
+    savingPermitted,
+    actions: {
+      saveAndReturn: () => {
+        if (savingPermitted && lastKnownDoc) {
+          onAppLeave((actions) => {
+            return actions.default();
+          });
```

Comment:
This needs a comment to explain what's going to happen, since you found it pretty unintuitive

Reply:
Added a quick comment to that.